test/integration/integration_test.cc

```diff
@@ -51,26 +51,45 @@ INSTANTIATE_TEST_SUITE_P(IpVersions, IntegrationTest,
                          TestUtility::ipTestParamsToString);
 
 // Make sure we have correctly specified per-worker performance stats.
-// TODO(mattklein123): We should flesh this test out to a) actually use more than 1 worker and
-// b) do some real requests and verify things work correctly on a per-worker basis. I will do this
-// in my next change when I add optional CX balancing as it well then be easier to write a
-// deterministic test.
-TEST_P(IntegrationTest, PerWorkerStats) {
+TEST_P(IntegrationTest, PerWorkerStatsAndBalancing) {
+  concurrency_ = 2;
+  config_helper_.addConfigModifier([&](envoy::config::bootstrap::v2::Bootstrap& bootstrap) -> void {
+    auto* listener = bootstrap.mutable_static_resources()->mutable_listeners(0);
+    listener->mutable_connection_balance_config()->mutable_exact_balance();
+  });
   initialize();
 
   // Per-worker listener stats.
-  if (GetParam() == Network::Address::IpVersion::v4) {
-    EXPECT_NE(nullptr, test_server_->counter("listener.127.0.0.1_0.worker_0.downstream_cx_total"));
-  } else {
-    EXPECT_NE(nullptr, test_server_->counter("listener.[__1]_0.worker_0.downstream_cx_total"));
-  }
+  auto check_listener_stats = [this](uint64_t expected_connections) {
+    if (GetParam() == Network::Address::IpVersion::v4) {
+      test_server_->waitForGaugeEq("listener.127.0.0.1_0.worker_0.downstream_cx_active",
+                                   expected_connections);
+      test_server_->waitForGaugeEq("listener.127.0.0.1_0.worker_1.downstream_cx_active",
+                                   expected_connections);
+    } else {
+      test_server_->waitForGaugeEq("listener.[__1]_0.worker_0.downstream_cx_active",
+                                   expected_connections);
+      test_server_->waitForGaugeEq("listener.[__1]_0.worker_1.downstream_cx_active",
+                                   expected_connections);
+    }
+  };
+  check_listener_stats(0);
 
   // Main thread admin listener stats.
   EXPECT_NE(nullptr, test_server_->counter("listener.admin.main_thread.downstream_cx_total"));
 
   // Per-thread watchdog stats.
   EXPECT_NE(nullptr, test_server_->counter("server.main_thread.watchdog_miss"));
   EXPECT_NE(nullptr, test_server_->counter("server.worker_0.watchdog_miss"));
+  EXPECT_NE(nullptr, test_server_->counter("server.worker_1.watchdog_miss"));
+
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  IntegrationCodecClientPtr codec_client2 = makeHttpConnection(lookupPort("http"));
+  check_listener_stats(1);
```

Comment:
Can you also check per-worker stats for downstream_cx_active?

Reply:
Yes I got lazy thanks for calling me out. :)

Reply:
I do check cx_active here so I assume you mean total? I will add that.